Source/ACE.Server/Managers/TransferManager.cs

```diff
@@ -0,0 +1,1152 @@
+            }
+
+            mre = new ManualResetEvent(false);
+            var slotCheck = false;
+            DatabaseManager.Shard.GetCharacters(metadata.AccountId, false, new Action<List<Character>>((chars) =>
+            {
+                slotCheck = chars.Count + 1 <= GameConfiguration.SlotCount;
+                mre.Set();
+            }));
+            mre.WaitOne();
+            if (!slotCheck)
+            {
+                return new ImportAndMigrateResult() { FailReason = ImportAndMigrateFailiureReason.NoCharacterSlotsAvailable };
+            }
+
+            TransferManagerCharacterMigrationDownloadResponseModel snapshotPack = null;
+            CharacterTransfer xfer = null;
+            if (importBytes == null)
+            {
+                mre = new ManualResetEvent(false);
+                DatabaseManager.Shard.GetCharacterTransfers((xfers) =>
+                {
+                    xfer = xfers.FirstOrDefault(k => k.Cookie == metadata.Cookie);
+                    mre.Set();
+                });
+                mre.WaitOne();
+                if (xfer != null)
+                {
+                    // don't fail here, prevents inter-account same-server transfers and name changes
+                    // return new ImportAndMigrateResult() { FailReason = ImportAndMigrateFailiureReason.CookieAlreadyUsed };
+                }
+
+                // try to verify we trust the server before downloading package, since migrations cause source server to delete the original char upon download of the package
+                string unverifiedThumbprintsSerialized = null;
+                try
+                {
+                    using (InsecureWebClient iwc = new InsecureWebClient())
+                    {
+                        unverifiedThumbprintsSerialized = iwc.DownloadString(metadata.ImportUrl.ToString() + "api/transferConfig");
+                    }
+                }
+                catch
+                {
+                    return new ImportAndMigrateResult() { FailReason = ImportAndMigrateFailiureReason.CannotContactSourceServer };
+                }
+                TransferManagerTransferConfigResponseModel unverifiedThumbprints = null;
+                try
+                {
+                    unverifiedThumbprints = JsonConvert.DeserializeObject<TransferManagerTransferConfigResponseModel>(unverifiedThumbprintsSerialized, serializationSettings);
+                }
+                catch
+                {
+                    return new ImportAndMigrateResult() { FailReason = ImportAndMigrateFailiureReason.ProtocolError };
+                }
+
+                if ((metadata.PackageType == PackageType.Migrate && !ConfigManager.Config.Transfer.AllowMigrationFrom.Any(k => k == unverifiedThumbprints.MyThumbprint)) || (metadata.PackageType == PackageType.Backup && !ConfigManager.Config.Transfer.AllowImportFrom.Any(k => k == unverifiedThumbprints.MyThumbprint)))
```

Comment:
this needs to move outside the "migration only" branch, and run the test for imports too